### PR TITLE
feat: created 'Circle' geometry component (#105)

### PR DESCRIPTION
**Pull request by ArielBenichou**
I have implemented a new React component for rendering circles on Google Maps within the existing library. Drawing inspiration from the `Marker` component, I made some code refinements, such as relocating the callbacks (e.g., `onClick`, `onDrag`) to a ref. This adjustment ensures that the `useEffect` does not need to be re-executed unnecessarily, addressing potential issues with un-memoized functions passed by the parent component. I encountered this issue involving an infinite feedback loop caused by this scenario.

Furthermore, I've included a sample application under `geometry` that showcases the circle component. This app allows users to interact with and manipulate the circle either through the map or via input controls.

I've also based the doc on the `Marker` markdown file.

This is my first time PR to an open-source project, so if i missed something, the feedback is welcome, thanks

### Changes Made

- Added a new React component for circles on Google Maps.
- Introduced a sample application (`geometry`) demonstrating circle manipulation via map interactions and input controls.
- Created documentation in the `docs/circle.md` file for easy reference.

### CODE REVIEW

**Comment by ArielBenichou** [comment]
## Initial Values
should we add an `initialRadius` and `initialCenter` to the `<Circle />` ?
imagine a case where you want to draw a Circle that is editable/draggable  but not controlled via a state.
```tsx
const [center, setCenter] = React.useState({lat: 41.23296, lng: -95.877363});
const [radius, setRadius] = React.useState(15000);

return (
  <>
    <Circle
        radius={radius}
        center={center}
        onRadiusChange={setRadius}
        onCenterChange={setCenter}
        draggable
        editable
     />
    <Circle 
        radius={10000}
        center={{lat: 41, lng: -95}}
        draggable
        editable
     />
  </>
)
```
in the case above, because we have one controlled component and one that is uncontrolled
the controlled component will trigger a re-render and reset the value of the uncontrolled component - snapping it back to it original position
using initials values for this will solve our problem.

**Comment by ArielBenichou** [comment]
I've also implemented the `<Polygon />` component, giving me this in total:
image

**Comment by usefulthink** [comment]
This is all very excellent stuff. I'm impressed by the overall quality of your work! I will definitely have a closer look at some point and adapt some of the patterns you used in the tests. I will add some notes to the code for some smaller issues I spotted.

The only thing I would like to see changed for now is to move all additions you made to the main source tree into the example folder. So `./src/components` would move to `./examples/geometry/src/components`, and `./docs/{circle,polygon}.md` would be moved to `./examples/geometry/docs/`. 

The lat/lng library functions can stay, but they could go together with the existing `is-lat-lng-literal.js` and become `lat-lng-utils.ts` or something (maybe also split everything, but I'm not a big fan of such small files...).

The reasoning for this is that we want to take additions to the main library a bit slower, and give new features some time to mature as part of the examples. This also gives us some time to think about if and how exactly we want to add them to the library ([I wrote a bite more about this here](https://visgl.github.io/react-google-maps/docs/contributing)).

In the end I want to prevent having to make changes to features after they are merged and published with the library, and thus trigger a major release. On the other hand, I don't see a problem with updating the code of an example if we discover a way to do things differently.

**Comment by ArielBenichou** [comment]
I've shuffled things around as you suggested - components and docs are now part of the "geometry" example.

I went ahead and bundled up the `lat-lng` utilities into one file (`lat-lng-utils.ts`). I'm on the same page of not wanting to split every function into its own thing.

About the tests, I scrapped them. They were pretty much a copy-pasta job from the marker tests with a few tweaks. Nothing too fancy.

And yeah, I'm totally get your approach of taking it slow before shoving stuff into the main package. Letting things marinate a bit sounds like a smart move.

Let's me know about anything else.

Cheers

File: examples/geometry/src/components/circle.tsx [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    onMouseOver,
+    onMouseOut,
+    onRadiusChanged,
+    onCenterChanged,
+    radius,
+    center,
+    ...circleOptions
+  } = props;
+  // This is here to avoid triggering the useEffect below when the callbacks change (which happen if the user didn't memoize them)
+  const callbacks = useRef<Record<string, (e: unknown) => void>>({});
```
Comment (usefulthink) [inline examples/geometry/src/components/circle.tsx:43]: That is indeed a nice way to handle this, I'll see if that pattern can be used for the map- and marker-events as well.

File: examples/geometry/src/components/circle.tsx [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    [
+      ['click', 'onClick'],
+      ['drag', 'onDrag'],
+      ['dragstart', 'onDragStart'],
+      ['dragend', 'onDragEnd'],
+      ['mouseover', 'onMouseOver'],
+      ['mouseout', 'onMouseOut']
+    ].forEach(([eventName, eventCallback]) => {
+      gme.addListener(circle, eventName, (e: google.maps.MapMouseEvent) =>
+        callbacks.current[eventCallback]?.(e)
```
Comment (usefulthink) [inline examples/geometry/src/components/circle.tsx:104]: I think I'd prefer a more explicit version here, something like 
```suggestion
        const callback = callbacks.current[eventCallback];
        if(callback) callback(e);
```

File: examples/geometry/src/components/polygon.tsx [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    [
+      ['click', 'onClick'],
+      ['drag', 'onDrag'],
+      ['dragstart', 'onDragStart'],
+      ['dragend', 'onDragEnd'],
+      ['mouseover', 'onMouseOver'],
+      ['mouseout', 'onMouseOut']
+    ].forEach(([eventName, eventCallback]) => {
+      gme.addListener(polygon, eventName, (e: google.maps.MapMouseEvent) =>
+        callbacks.current[eventCallback]?.(e)
```
Comment (usefulthink) [inline examples/geometry/src/components/polygon.tsx:111]: see above.

**Review by usefulthink — CHANGES_REQUESTED** [review]
Excellent, thanks. A couple more small notes, then we can get this merged.

File: docs/api-reference/components/polygon.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    <Map zoom={12} center={{lat: 53.54992, lng: 10.00678}}>
+      {/* Draw the Bermuda triangle */}
+      <Polygon paths={[
+        { lat: 25.774, lng: -80.190},
+        { lat: 18.466, lng: -66.118},
+        { lat: 32.321, lng: -64.757},
+      ]} />
+
+      {/* Draw the Bermuda triangle with an encoded path */}
+      <Polygon encodedPaths=["o~h|CnbmhN~irk@_m{tAw`qsAgyhG"] />
```
[the remark this answers is not in the document]
Comment (ArielBenichou) [reply docs/api-reference/components/polygon.md]: yes this is a typo, but the `encodedPaths` get an array of paths.
```ts
  encodedPaths?: string[];
```
so it should be: 
```tsx
<Polygon encodedPaths={["o~h|CnbmhN~irk@_m{tAw`qsAgyhG"]} />
```

File: examples/geometry/src/components/circle.tsx (hunk 1 above)
Comment (ArielBenichou) [reply examples/geometry/src/components/circle.tsx:43]: 😄

File: examples/geometry/src/polygons.ts
[the remark this answers is not in the document]
Comment (ArielBenichou) [reply examples/geometry/src/polygons.ts:1]: yeah that a big scary file, TBH i took this from the project i'm working on, maybe it is better to simplify it to a simple triangle?
I just thought it looked cool XD

File: examples/geometry/README.md
[the remark this answers is not in the document]
Comment (ArielBenichou) [reply examples/geometry/README.md:1]: yeah, that's a great idea

File: examples/geometry/src/polygons.ts
[the remark this answers is not in the document]
Comment (usefulthink) [reply examples/geometry/src/polygons.ts:1]: No, thats perfectly fine, examples should be looking interesting and useful :)

I think it just needs a comment that those are polygons of the districts of Omaha, Nebraska (?) in the encoded polyline algorithm format, with a link to this: https://developers.google.com/maps/documentation/utilities/polylinealgorithm and maybe a link to some tool you used to generate them.

**Review by usefulthink — APPROVED** [review]

**Comment by usefulthink** [comment]
Again, thanks a lot. Last thing to do is to update the website files to add the example to the overview page, do you want to do that? The steps for that are described here: https://visgl.github.io/react-google-maps/docs/guides/writing-examples#adding-examples-to-the-website

**Comment by ArielBenichou** [comment]
No problem, let's do things all the way.
I will create a new PR for the website example.

**Pull request merged**